src/sparse/csmat.rs

```diff
@@ -1518,6 +1527,32 @@ where
             }
         })
     }
+
+    /// Iteration over all entries on the diagonal
+    pub fn diag_iter_mut(
+        &mut self,
+    ) -> impl ExactSizeIterator<Item = Option<&mut N>>
+           + DoubleEndedIterator<Item = Option<&mut N>>
+           + '_ {
+        let data_ptr: *mut N = self.data[..].as_mut_ptr();
+        let smallest_dim = cmp::min(self.ncols, self.nrows);
+        (0..smallest_dim).map(move |i| {
+            let idx = self.nnz_index_outer_inner(i, i);
+            if let Some(NnzIndex(idx)) = idx {
+                // To obtain multiple mutable references to different
+                // locations in data we must use a pointer and some unsafe.
+                // # Safety
+                // This is safe as
+                // * NnzIndex provides bounds checking
+                // * diagonal entries are never overlapping in memory
+                // * no entries are requested more than once
+                // * nnz_index_outer_inner does not modify or read from entries in self.data
+                Some(unsafe { &mut *data_ptr.add(idx) })
```

Comment:
Quick thought, I haven't checked yet if it works: the nnz indices of the diagonal elements should be increasing, so maybe it's possible to keep using safe rust only by using `slice::split_at_mut` and remembering the last nnz index to have the correct offset?

Or we could have an iterator that yields mutable outer views, and chaining that we'd be sure we have at most one reference per mutable outer view.

Reply:
I was thinking of iterating through the outer views, but could not work out the lifetimes, Could be this is one of the places where we need a reborrowing outer view iterator

Reply:
Actually I'm not sure it's possible to use the reborrowing pattern with mutable data as it would break the single reference. So it's possible there's no way other than reimplementing the whole logic for each iterator. Your unsafe impl might be the easiest way actually.

Reply:
`outer_iterator_mut_rbr` seems to be impossible to create using the current implementation of `outer_iterator`. I noticed `outer_iterator` implements `DoubleEndedIterator` but only works forward using repeated `split_at_mut`. I don't think this is valid?

Reply:
You're right, I think I assumed I could do a `DoubleEndedIterator` as in the immutable case, but the current implementation cannot fulfill that bound. I'm not even sure it works in the forward case now, I translated the old implementation to the `impl Iterator` one but as this is not tested a lot I may have introduced issues.

Reply:
It's probably possible to implement the `DoubleEndedIterator` by keeping around a range that indicates which part of the original data the current range corresponds to.